cli/azd/docs/external-prompting.md

```diff
@@ -0,0 +1,114 @@
+
+`${AZD_UI_PROMPT_ENDPOINT}/prompt?api-version=2024-02-14-preview`
+
+Setting the following headers:
+
+- `Content-Type: application/json`
+- `Authorization: Bearer ${AZD_UI_PROMPT_KEY}`
+
+The use of `AZD_UI_PROMPT_KEY` allows the host to block requests coming from other clients on the same machine (since the it is expected the host runs a ephemeral HTTP server listing on `127.0.0.1` on a random port). It is expected that the host will generate a random string and use this as a shared key for the lifetime of an `azd` invocation.
+
+The body of the request contains information about the prompt that `azd` needs a response for:
+
+```jsonc
+{
+    "type": "<string>", // one of "string", "password", "select", "multiSelect", "confirm",
+    "options": {
+      "message": "<string>", // the message to be displayed as part of the prompt
+      "help": "<string>", // optional help text that can be displayed upon request
+      "options": [ "<string>" /* ... */  ], // for select and multiSelect types - the valid options the user should pick from
+      "defaultValue": "<string>", // optional default value, when undefined there is no default.
+    }
+}
+```
+
+The `password` type represents a string value which represents a password. The host may want to use a different UI element (perhaps one that uses `***` instead of characters) when prompting.
+
+The server should respond with 200 OK and the body should be one of the of the following three shapes:
+
+### Success 
+
+When the host is able to prompt the value it is returned with a `result` of `success`:
+
+```jsonc
+{
+    "result": "success", 
+    "value": "<string> | <string>[]"
+}
+```
+
+When the type is `confirm` the value should be either `"true"` or `"false"` (a string, not a JSON boolean) indicating if the user confirmed the operation (`"true"`) or rejected it (`"false"`). Note that a user rejecting a confirm prompt still results in a `"success"` result (the value is simply `"false"`). In the case of `multiSelect` an array of string values is returned, each individual value is a value from the `"options"` array that was selected by the user.
+
+### Cancelled
+
+The user may decline to provide a response (imagine hitting a cancel button on the dialog that is being used to present the question, or a user hitting something like CTRL+C in a terminal interaction to abort the question asking but not the entire application).
+
+In this case the `result` is `cancelled`.
+
+```jsonc
+{
+    "result": "cancelled", 
+}
+```
+
+`azd` returns a special `Error` type internally in this case, which up-stack code can use.
+
+### Error
+
+Some error happened during prompting - the error message will be returned as is.
+
+```jsonc
+{   "result": "error",
+    "error": {
+        "message": "<string>" // an error message, will be the text of the `Error` returned inside `azd`.
```

Comment:
Interesting... I guess the user interacting with `azd` through a terminal?

Reply:
This is more about providing the caller a way to control the message in the `error` that we will return from the failed prompt call.  What is done with that message depends on how the caller of the prompt handles errors.  But yes, in cases like VS Code where the output of `azd` is wired to the interactive terminal this message would be rendered at the end of the Error message when the `azd` operation fails.